{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello world\n",
      "Hello world\n",
      "She said, 'Hello world'\n",
      "Comments are very useful!\n",
      "Hello world\n",
      "Hello\n",
      "world\n",
      "ATTGCTTC\n",
      "Helloworld\n",
      "12\n",
      "The length of the DNA sequence is9\n",
      "atggct\n",
      "clsgpadyn\n",
      "vslgpadvn\n",
      "gp\n",
      "clsgpa\n",
      "clsgpadvn\n",
      "s\n",
      "valines:2\n",
      "tryptophans:0\n",
      "0\n",
      "-1\n"
     ]
    }
   ],
   "source": [
    "print(\"Hello world\")\n",
    "print('Hello world')\n",
    "print(\"She said, 'Hello world'\")\n",
    "\n",
    "#This is a comment; it will be ignored by the computer\n",
    "print(\"Comments are very useful!\")\n",
    "\n",
    "#Print a friendly greeting\n",
    "print(\"Hello world\")\n",
    "\n",
    "#Print is a function-tells python what we want to do\n",
    "#Text appearing in parentheses are arguments to the function-gives more specific instructions to python\n",
    "#When using more than one argument, the arguments should be separated by commas.\n",
    "\n",
    "#Strings are always surrounded by quotes. This allows python to distinguish between instructions and data\n",
    "#Output will not contain the quotes because they are part of the code (You can have quotes in output though. See above)\n",
    "\n",
    "#Failure to put the argument in quotes will yield a syntax error\n",
    "#print(Hello world)\n",
    "\n",
    "#Misspelling the the name of a funciton returns a name error\n",
    "#prn(\"Hello world\")\n",
    "\n",
    "#You cannot split a statement over two lines by hitting enter\n",
    "#print(\"Hello\n",
    "#world\")\n",
    "\n",
    "#How to include a newline in the middle of a string\n",
    "print(\"Hello\\nworld\")\n",
    "#Do not include a space either before or after the newline character\n",
    "\n",
    "#Store a short DNA sequence in a variable\n",
    "my_dna = \"ATTGCTTC\"\n",
    "\n",
    "#Print the DNA sequence\n",
    "print(my_dna)\n",
    "#When you use a variable in a function you don't have to include the quotes because they are already part of the variable\n",
    "\n",
    "#You can change the value of a variable at any point\n",
    "\n",
    "#Change the value of my_dna\n",
    "my_dna = \"GGCTGTAA\"\n",
    "\n",
    "#Store a short sequence in the variable name banana\n",
    "banana = \"GATCTAC\"\n",
    "\n",
    "#Variable names are arbitrary, but try to use meaningful names (banana is a bad name)\n",
    "\n",
    "#Concatenation-add two strings together using +\n",
    "my_dna = \"AATG\" + \"TGCGTA\"\n",
    "\n",
    "#Other examples of concatenation\n",
    "upstream = \"TTTTT\"\n",
    "my_dna = upstream + \"ATGCA\"\n",
    "\n",
    "upstream = \"ATG\"\n",
    "downstream = \"AAA\"\n",
    "my_dna = upstream + \"ATGCGT\" + downstream\n",
    "\n",
    "#Use concatenation inside a print statement\n",
    "print(\"Hello\"+\"world\") #will not include a space between hello and world#\n",
    "\n",
    "#Find the length of a string (does not produce visual output)\n",
    "len(\"ATCG\")\n",
    "\n",
    "#To produce output with len function, store output as a variable and print it (output is an integer)\n",
    "dna_length = len(my_dna)\n",
    "print(dna_length)\n",
    "\n",
    "#You cannot concatenate strings and integers because they are different types\n",
    "#my_dna = \"ATGCGTGATG\"\n",
    "#dna_length = len(my_dna)\n",
    "#Print(\"The length of the DNA sequence is\" + dna_length) -- will produce error\n",
    "\n",
    "#Types are different kinds of information that can exist in a program\n",
    "\n",
    "#There are functions for converting types\n",
    "#To convert a type to a string use the str function\n",
    "\n",
    "#To properly calculate the length of a DNA sequence\n",
    "my_dna = \"ATGCTCTAA\"\n",
    "dna_length = len(my_dna)\n",
    "print(\"The length of the DNA sequence is\" + str(dna_length))\n",
    "#For every parentheses you open you have to close\n",
    "\n",
    "#To change case\n",
    "    #You can convert a string to lowercase or uppercase using a method\n",
    "    #A method is similar to a function, but they are particular to types\n",
    "    #EX: the lower method is unique to strings (same is true of upper method)\n",
    "    #Format = variable.nameofmethod(arguments)\n",
    "    \n",
    "my_dna = \"ATGGCT\"\n",
    "#Print my_dna in lower case\n",
    "print(my_dna.lower())\n",
    "\n",
    "#lower and upper methods do not actually change the value of the variable\n",
    "\n",
    "#replace = another method - used to find and replace parts of strings\n",
    "    #Takes two arguments and returns a copy of the variable where all occurances of the 1st string are replaced\n",
    "    #by the second string\n",
    "\n",
    "protein = \"clsgpadvn\"\n",
    "\n",
    "#replace valine with tyrosine\n",
    "print(protein.replace(\"v\", \"y\"))\n",
    "\n",
    "#replace more than one character\n",
    "print(protein.replace(\"cls\", \"vsl\"))\n",
    "#again, this function does not actually alter the variable\n",
    "\n",
    "#Extracting part of a string\n",
    "    #substring = part of a string\n",
    "    #follow variable name with square brackets enclosing a start and stop position separated by :\n",
    "    #Start position is inclusive; stop position is exclusive\n",
    "    #Numbering starts at 0 not 1\n",
    "    #If you use a stop position well beyond the end of the string it will print to the end\n",
    "    #If you put a single # in the square brackets you get a single character\n",
    "\n",
    "#Print positions 3 through five of protein\n",
    "print(protein[3:5])\n",
    "print(protein[0:6])\n",
    "print(protein[0:75])\n",
    "print(protein[2])\n",
    "\n",
    "#Counting and finding substrings\n",
    "    #use count method to count #of times a substring occurs in a string\n",
    "        #return type is integer\n",
    "    #use find method to find location of substring in a string\n",
    "        #returns a number referred to as the index\n",
    "        #if you ask it to locate a substring that doesn't exist -1 will be output\n",
    "    #count and find can only be used to locate exact matches\n",
    "\n",
    "protein = \"vlspadktnv\"\n",
    "#Count the number of valine amino acid residues\n",
    "valine_count = protein.count('v')\n",
    "#Count the number of tryptophans\n",
    "tryptophan_count = protein.count('w')\n",
    "#Print the counts\n",
    "print('valines:'+ str(valine_count))\n",
    "print('tryptophans:' + str(tryptophan_count))\n",
    "#Locate valine residues\n",
    "print(str(protein.find('v')))\n",
    "#Locate the tryptophan residues\n",
    "print(str(protein.find('w')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6851851851851852\n"
     ]
    }
   ],
   "source": [
    "#Exercise 1-Calculating AT content\n",
    "\n",
    "#Calculate the AT content of the DNA sequence below\n",
    "#ACTGATCGATTACGTATAGTATTTGCTATCATACATATATATCGATGCGTTCAT\n",
    "\n",
    "dna_seq = \"ACTGATCGATTACGTATAGTATTTGCTATCATACATATATATCGATGCGTTCAT\"\n",
    "\n",
    "#Count the occurances of 'A'\n",
    "a_content = dna_seq.count('A')\n",
    "\n",
    "#Count the occurances of 'T'\n",
    "t_content = dna_seq.count('T')\n",
    "\n",
    "#Add occurances of 'A' to occurances of 'T'\n",
    "at_together = a_content + t_content\n",
    "\n",
    "#Find the length of the DNA sequence\n",
    "dna_length = len(dna_seq)\n",
    "\n",
    "#Divide 'A + T' by length of sequence\n",
    "at_content = at_together/dna_length\n",
    "print(at_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGACTAGCTAATGCATATCATAAACGATAGTATGTATATATAGCTACGCAAGTA\n"
     ]
    }
   ],
   "source": [
    "#Exercise 2-Complementing DNA\n",
    "\n",
    "#Write a program that will print the complement of the DNA sequence below\n",
    "#ACTGATCGATTACGTATAGTATTTGCTATCATACATATATATCGATGCGTTCAT\n",
    "\n",
    "dna_seq2 = \"ACTGATCGATTACGTATAGTATTTGCTATCATACATATATATCGATGCGTTCAT\"\n",
    "\n",
    "#Replace 'A' with 't'\n",
    "new_seq = dna_seq2.replace('A', 't')\n",
    "\n",
    "#Replace 'C' with 'G'\n",
    "new_seq2 = new_seq.replace('C', 'g')\n",
    "\n",
    "#Replace 'T' with 'A'\n",
    "new_seq3 = new_seq2.replace('T', 'a')\n",
    "\n",
    "#Replace 'G' with 'C'\n",
    "fin_seq = new_seq3.replace('G', 'c')\n",
    "\n",
    "#Print the final sequence (Note: printed 3' to 5')\n",
    "print(fin_seq.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of the DNA sequence is:55\n",
      "The index of the recognition site is:21\n",
      "The sequence of the first fragementment is:ACTGATCGATTACGTATAGTAG\n",
      "The sequence of the second fragment is:AATTCTATCATACATATATATCGATGCGTTCAT\n",
      "The length of the first fragement is:22\n",
      "The length of the second fragment is:33\n",
      "The length of the two fragments should equal the length of the total sequence\n",
      "The length of the two fragments together is:55\n",
      "length of fragment 1 is:22\n",
      "length of fragment 2 is:33\n"
     ]
    }
   ],
   "source": [
    "#Exercise 3-Restriction fragment lengths\n",
    "\n",
    "#The sequence below contains a recognition site for the EcoRI restriction enzyme which cuts at the motif G*AATTC\n",
    "#Write a program that will calculate the size of the two fragments it will produce from the following DNA sequence\n",
    "#ACTGATCGATTACGTATAGTAGAATTCTATCATACATATATATCGATGCGTTCAT\n",
    "\n",
    "dna_seq = \"ACTGATCGATTACGTATAGTAGAATTCTATCATACATATATATCGATGCGTTCAT\"\n",
    "\n",
    "#Find the length of the sequence\n",
    "length = len(dna_seq)\n",
    "print('The length of the DNA sequence is:' + str(length))\n",
    "\n",
    "#Find the recognition site\n",
    "rec_seq = 'GAATTC'\n",
    "rec_site = dna_seq.find(rec_seq)\n",
    "print('The index of the recognition site is:' + str(rec_site))\n",
    "\n",
    "#Split the string\n",
    "frag1 = dna_seq[0:rec_site+1]\n",
    "print('The sequence of the first fragementment is:' + frag1)\n",
    "\n",
    "frag2 = dna_seq[rec_site+1:55]\n",
    "print('The sequence of the second fragment is:' + frag2)\n",
    "\n",
    "#Calculate the lengths of the two resulting strings\n",
    "print('The length of the first fragement is:' + str(len(frag1)))\n",
    "print('The length of the second fragment is:' + str(len(frag2)))\n",
    "\n",
    "#Double check that fragment sizes make sense\n",
    "print('The length of the two fragments should equal the length of the total sequence')\n",
    "print('The length of the two fragments together is:' + str(len(frag1)+len(frag2)))\n",
    "\n",
    "#Solution from textbook\n",
    "frag1_length = dna_seq.find('GAATTC') + 1\n",
    "frag2_length = len(dna_seq) - frag1_length\n",
    "print('length of fragment 1 is:' + str(frag1_length))\n",
    "print('length of fragment 2 is:' + str(frag2_length))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my output\n",
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCATCGATCGATATCGATGCATCGACTACTAT\n",
      "Output from text\n",
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCATCGATCGATATCGATGCATCGACTACTAT\n",
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCATCGATCGATATCGATGCATCGACTACTAT\n",
      "my exon 1\n",
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGA\n",
      "text exon 1\n",
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCG\n",
      "my exon 2\n",
      "TCATCGATCGATATCGATGCATCGACTACTAT\n",
      "text exon 2\n",
      "ATCATCGATCGATATCGATGCATCGACTACTAT\n"
     ]
    }
   ],
   "source": [
    "#Exercise 4-Splicing out introns part 1\n",
    "\n",
    "#The DNA sequence below contains two exons and an intron\n",
    "#Exon 1 start to 63rd character\n",
    "#Exon 2 91st character to end\n",
    "\n",
    "#Write a program that will print just the coding regions of the DNA sequence\n",
    "\n",
    "#ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCGATCGATCGATCGATCGATCATGCTATCATCGATCGATATCGATGCATCGACTACTAT\n",
    "\n",
    "dna_seq = \"ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCGATCGATCGATCGATCGATCATGCTATCATCGATCGATATCGATGCATCGACTACTAT\"\n",
    "\n",
    "print('my output')\n",
    "print(dna_seq[0:63] + dna_seq[91:500])\n",
    "\n",
    "#Check my output with output from textbook\n",
    "print('Output from text')\n",
    "print('ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCATCGATCGATATCGATGCATCGACTACTAT')\n",
    "\n",
    "#Code from textbook\n",
    "exon1 = dna_seq[0:62]\n",
    "exon2 = dna_seq[90:500]\n",
    "print(exon1 + exon2)\n",
    "\n",
    "#Why am I using different characters and getting the same output as the textbook?\n",
    "print('my exon 1')\n",
    "print(dna_seq[0:63])\n",
    "\n",
    "print('text exon 1')\n",
    "print(exon1)\n",
    "\n",
    "print('my exon 2')\n",
    "print(dna_seq[91:500])\n",
    "\n",
    "print('text exon 2')\n",
    "print(exon2)\n",
    "\n",
    "#The text states that exon 1 begins at (and includes) the first character (which to python is the 0 character)\n",
    "#Exon 1 ends at (but does NOT include) the 63rd character (which to python is the 62nd character)\n",
    "#Exon 2 starts at (and includes) the 91st character (which to python is the 90th character)\n",
    "#My code for exon1 includes the 1st character of the intron which is an A\n",
    "#My code for exon2 excludes the first character of exon2 which is also an A\n",
    "#Due to the fact that the first character of the intron and exon2 are the same, my mistake cancelled out\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "123\n",
      "95\n",
      "77.23577235772358\n"
     ]
    }
   ],
   "source": [
    "#Exercise 5-Splicing out introns part 2\n",
    "\n",
    "#Using the data from part 1, write a program to calculate % of sequence that is coding\n",
    "\n",
    "#Calculate length of entire sequence\n",
    "print(len(dna_seq))\n",
    "\n",
    "#Calculate length of coding region\n",
    "coding_reg = exon1 + exon2\n",
    "print(len(coding_reg))\n",
    "\n",
    "#length of exons divided by length of sequence\n",
    "ratio = len(coding_reg)/len(dna_seq)\n",
    "\n",
    "#Multiply by 100\n",
    "print(str(ratio*100))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGatcgatcgatcgatcgatcgatcatgctATCATCGATCGATATCGATGCATCGACTACTAT\n",
      "text output\n",
      "ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGatcgatcgatcgatcgatcgatcatgctATCATCGATCGATATCGATGCATCGACTACTAT\n"
     ]
    }
   ],
   "source": [
    "#Exercise 6-Splicing out introns part 3\n",
    "\n",
    "#Write a program that will print the original sequence with intron in lowercase\n",
    "\n",
    "intron = dna_seq[62:90]\n",
    "\n",
    "print(exon1 + intron.lower() + exon2)\n",
    "\n",
    "#Compare my output to text output\n",
    "print('text output')\n",
    "print('ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGatcgatcgatcgatcgatcgatcatgctATCATCGATCGATATCGATGCATCGACTACTAT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
